pyqtorch/core/batched_operation.py

```diff
@@ -20,12 +20,54 @@
 
 from pyqtorch.converters.store_ops import ops_cache, store_operation
 from pyqtorch.core.operation import RX, H
-from pyqtorch.core.utils import _apply_batch_gate
+from pyqtorch.core.utils import _apply_batch_gate, OPERATIONS_DICT
 
-IMAT = torch.eye(2, dtype=torch.cdouble)
-XMAT = torch.tensor([[0, 1], [1, 0]], dtype=torch.cdouble)
-YMAT = torch.tensor([[0, -1j], [1j, 0]], dtype=torch.cdouble)
-ZMAT = torch.tensor([[1, 0], [0, -1]], dtype=torch.cdouble)
+IMAT = OPERATIONS_DICT["I"]
+XMAT = OPERATIONS_DICT["RX"]
+YMAT = OPERATIONS_DICT["RY"]
+ZMAT = OPERATIONS_DICT["RZ"]
+
+
+def get_parametrized_batch_for_operation(operation_type: str, theta: torch.Tensor, batch_size: int, device: torch.device) -> torch.Tensor:
+    """ Helper method which takes a string describing an operation type and a parameter theta vector and returns
+        a batch of the corresponding parametrized rotation matrices
+    Args:
+
+    operation_type (str): the type of operation which should be performed (RX,RY,RZ)
+    theta (torch.Tensor): 1D-tensor holding the values of the parameter
+    batch_size (int): the batch size
+    device (torch.device): the device which to run on
+
+    Returns:
+    torch.Tensor: a batch of gates after applying theta 
+    """
+
+    cos_t = torch.cos(theta / 2).unsqueeze(0).unsqueeze(1)
+    cos_t = cos_t.repeat((2, 2, 1))
+    sin_t = torch.sin(theta / 2).unsqueeze(0).unsqueeze(1)
+    sin_t = sin_t.repeat((2, 2, 1))
+
+    batch_imat = OPERATIONS_DICT["I"].unsqueeze(2).repeat(1, 1, batch_size).to(device)
+    batch_operation_mat = OPERATIONS_DICT[operation_type].unsqueeze(2).repeat(1, 1, batch_size).to(device)
+
+    return cos_t * batch_imat - 1j * sin_t * batch_operation_mat
+
+
+def create_controlled_batch_from_operation(operation_batch: torch.Tensor, batch_size: int) -> torch.Tensor:
+    """ Method which takes a 2x2 torch.Tensor and transforms it into a Controlled Operation Gate
+
+    Args:
+
+        operation_matrix (torch.Tensor): the type of operation which should be performed (RX,RY,RZ)
+        batch_size (int): the batch size
+    
+    Returns:
+
+        torch.Tensor: the resulting controlled gate populated by operation_matrix
+    """
+    controlled_batch: torch.Tensor = torch.eye(4, dtype=torch.cdouble).unsqueeze(2).repeat(1, 1, batch_size)
+    controlled_batch[2:,2:,:] = operation_batch
```

Comment:
I think it would be safer to add `controlled_batch[2:,2:,:] = operation_batch.copy()`

Reply:
done